internal/core/docker_context.go

```diff
@@ -0,0 +1,328 @@
+	Endpoints map[string]json.RawMessage `json:"endpoints,omitempty"`
+	Name      string                     `json:"name,omitempty"`
+}
+
+func (s *metadataStore) getByID(id contextdir) (metadata, error) {
+	fileName := filepath.Join(s.contextDir(id), metaFile)
+	bytes, err := os.ReadFile(fileName)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return metadata{}, errdefs.NotFound(fmt.Errorf("context not found: %w", err))
+		}
+		return metadata{}, err
+	}
+
+	var untyped untypedContextMetadata
+	r := metadata{
+		Endpoints: make(map[string]any),
+	}
+
+	if err := json.Unmarshal(bytes, &untyped); err != nil {
+		return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
+	}
+
+	r.Name = untyped.Name
+	if r.Metadata, err = parseTypedOrMap(untyped.Metadata, s.config.contextType); err != nil {
+		return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
+	}
+
+	for k, v := range untyped.Endpoints {
+		if r.Endpoints[k], err = parseTypedOrMap(v, s.config.endpointTypes[k]); err != nil {
+			return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
+		}
+	}
+
+	return r, err
+}
+
+// list returns a list of all Docker contexts
+func (s *metadataStore) list() ([]metadata, error) {
+	ctxDirs, err := listRecursivelyMetadataDirs(s.root)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	res := make([]metadata, 0, len(ctxDirs))
+	for _, dir := range ctxDirs {
+		c, err := s.getByID(contextdir(dir))
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return nil, fmt.Errorf("read metadata: %w", err)
+		}
+		res = append(res, c)
+	}
+
+	return res, nil
+}
+
+type contextdir string
+
+func isContextDir(path string) bool {
+	s, err := os.Stat(filepath.Join(path, metaFile))
+	if err != nil {
+		return false
+	}
+
+	return !s.IsDir()
+}
+
+func listRecursivelyMetadataDirs(root string) ([]string, error) {
+	fis, err := os.ReadDir(root)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []string
+	for _, fi := range fis {
+		if fi.IsDir() {
+			if isContextDir(filepath.Join(root, fi.Name())) {
+				result = append(result, fi.Name())
+			}
+
+			subs, err := listRecursivelyMetadataDirs(filepath.Join(root, fi.Name()))
+			if err != nil {
+				return nil, err
+			}
+
+			for _, s := range subs {
+				result = append(result, filepath.Join(fi.Name(), s))
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func parseTypedOrMap(payload []byte, getter typeGetter) (any, error) {
+	if len(payload) == 0 || string(payload) == "null" {
+		return nil, nil
+	}
+
+	if getter == nil {
+		var res map[string]any
+		if err := json.Unmarshal(payload, &res); err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+
+	typed := getter()
+	if err := json.Unmarshal(payload, typed); err != nil {
+		return nil, err
+	}
+
+	return reflect.ValueOf(typed).Elem().Interface(), nil
+}
+
+// getHomeDir returns the home directory of the current user with the help of
```

Comment:
suggestion: replace with `os.UserHomeDir`, I suspect this is copy from a time when that didn't exist.

Reply:
It's using `os.UserHomeDir` already, with a corner case where the home is empty in not-windows 🤔 
```go
// [pkg/homedir.Get]: https://pkg.go.dev/github.com/docker/docker@v26.1.4+incompatible/pkg/homedir#Get
func getHomeDir() string {
	home, _ := os.UserHomeDir()
	if home == "" && runtime.GOOS != "windows" {
		if u, err := user.Current(); err == nil {
			return u.HomeDir
		}
	}
	return home
}
```

Reply:
I'll open my eyes next time 🤣